### PR TITLE
Optimize the speed of the truncate_sequences function.

### PR DESCRIPTION
**Pull request by ikkvix**
# What does this PR do?

The truncate_sequences function uses a for loop to iteratively remove elements when truncation_strategy == TruncationStrategy.LONGEST_FIRST. However, when the length of ids or pair_ids is considerable, the time complexity of the removal process becomes O(n^2), which is unacceptable. By directly calculating the quantity to be removed, this portion of the code's time complexity can be reduced to O(n), resulting in a significant speed improvement.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by ikkvix** [comment]
@ArthurZucker  @younesbelkada  Could you give me any opinions or suggestions?

**Comment by younesbelkada** [comment]
Hi @ikkvix thanks for your great work on this, I will let @ArthurZucker review the PR as he is much more familiar on tokenizers that I am

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -3557,21 +3557,26 @@ def truncate_sequences(
...
+            len_pair_ids = len(pair_ids) if pair_ids is not None else 0
+            len_ids = len(ids)
+            first_remove = min(abs(len_pair_ids - len_ids), num_tokens_to_remove)
+            second_remove = num_tokens_to_remove - first_remove
+            if len_ids > len_pair_ids:
+                ids_to_move = first_remove + second_remove // 2
+                pair_ids_to_move = second_remove - second_remove // 2
+            else:
+                ids_to_move = second_remove // 2
+                pair_ids_to_move = first_remove + (second_remove - second_remove // 2)
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:3569]:
```suggestion
                pair_ids_to_move = first_remove + second_remove - (second_remove // 2)
```

**Review by ArthurZucker — APPROVED** [review]
LGTM, what kind of speedups are we looking at? (did you benchmark it?)

**Comment by ikkvix** [comment]
> LGTM, what kind of speedups are we looking at? (did you benchmark it?)

Thank you for the review! I conducted some experiments on my pc:

- When num_tokens_to_remove=10,000:
    The old method takes 0.2s in this section of the code, while the new method takes 1e-05s.

- When num_tokens_to_remove=100,000:
    The old method takes 18.94s in this section of the code, while the new method only takes 1.38e-05s.

- When num_tokens_to_remove=200,000:
    The old method takes 75.21s in this section of the code, while the new method only takes 1.81e-05s.

- When num_tokens_to_remove=300,000:
    The old method takes 176.97s in this section of the code, while the new method takes 2.10e-05s.

I didn't measure at the million level because it took too long. It's evident that with the increase in num_tokens_to_remove, the speed of the for loop removal becomes extremely slow, highlighting a massive difference in speed between these two methods.

**Comment by ArthurZucker** [comment]
Thanks for improving the truncation. This is mostly used in `prepare_for_model`  which is used in `encode` basically. A lot of model use this with `truncation=True`. Let me just run the slow tokenization tests to make sure everything is alright

**Comment by ArthurZucker** [comment]
No new failures in the slow tokenization tests, checking the `RUN_CUSTOM_TOKENIZERS` (most of them use truncate_sequence individually) and seems alright! 🪂 thanks

**Pull request merged**